sys/stdvga.c

```diff
@@ -0,0 +1,165 @@
+#define VGA_QEMU_STDVGA_VENDOR_ID 0x1234
+#define VGA_QEMU_STDVGA_DEVICE_ID 0x1111
+
+static void stdvga_io_write(stdvga_device_t *vga, uint16_t reg, uint8_t value) {
+  bus_space_write_1(&vga->io, reg + VGA_MMIO_OFFSET, value);
+}
+static uint8_t stdvga_io_read(stdvga_device_t *vga, uint16_t reg)
+  __attribute__((unused));
+static uint8_t stdvga_io_read(stdvga_device_t *vga, uint16_t reg) {
+  return bus_space_read_1(&vga->io, reg + VGA_MMIO_OFFSET);
+}
+static void stdvga_vbe_write(stdvga_device_t *vga, uint16_t reg,
+                             uint16_t value) {
+  /* <<1 shift enables access to 16-bit registers. */
+  bus_space_write_2(&vga->io, (reg << 1) + VBE_MMIO_OFFSET, value);
+}
+static uint16_t stdvga_vbe_read(stdvga_device_t *vga, uint16_t reg) {
+  return bus_space_read_2(&vga->io, (reg << 1) + VBE_MMIO_OFFSET);
+}
+
+static void stdvga_palette_write_single(stdvga_device_t *stdvga, uint8_t offset,
+                                        uint8_t r, uint8_t g, uint8_t b) {
+  stdvga_io_write(stdvga, VGA_PALETTE_ADDR, offset);
+  stdvga_io_write(stdvga, VGA_PALETTE_DATA, r >> 2);
+  stdvga_io_write(stdvga, VGA_PALETTE_DATA, g >> 2);
+  stdvga_io_write(stdvga, VGA_PALETTE_DATA, b >> 2);
+}
+
+static void stdvga_palette_write_buffer(stdvga_device_t *stdvga,
+                                        const uint8_t buf[VGA_PALETTE_SIZE]) {
+  for (int i = 0; i < VGA_PALETTE_SIZE / 3; i++)
+    stdvga_palette_write_single(stdvga, i, buf[3 * i + 0], buf[3 * i + 1],
+                                buf[3 * i + 2]);
+}
+
+static int stdvga_palette_write(vga_device_t *vga, uio_t *uio) {
+  stdvga_device_t *stdvga = STDVGA_FROM_VGA(vga);
+  int error = uiomove_frombuf(stdvga->palette_buffer, VGA_PALETTE_SIZE, uio);
+  if (error)
+    return error;
+  /* TODO: Only update the modified area. */
+  stdvga_palette_write_buffer(stdvga, stdvga->palette_buffer);
+  return 0;
+}
+
+static int stdvga_fb_write(vga_device_t *vga, uio_t *uio) {
+  stdvga_device_t *stdvga = STDVGA_FROM_VGA(vga);
+  /* TODO: I'd love to pass the uio to the bus directly, instead of using a
+     buffer to write out bytes one by one... */
+  int error =
+    uiomove_frombuf(stdvga->fb_buffer, stdvga->width * stdvga->height, uio);
+  if (error)
+    return error;
+  bus_space_write_region_1(&stdvga->mem, 0, stdvga->fb_buffer,
+                           stdvga->width * stdvga->height);
+  return 0;
+}
+
+MALLOC_DEFINE(stdvga_pool, "stdvga VGA driver pool");
+
+int stdvga_pci_attach(pci_device_t *pci) {
+  if (pci->vendor_id != VGA_QEMU_STDVGA_VENDOR_ID ||
+      pci->device_id != VGA_QEMU_STDVGA_DEVICE_ID)
+    return 0;
+
+  /* TODO: Initialize the pool somewhere else... This assumes "attach" will only
+     get called once. */
+  kmalloc_init(stdvga_pool, 128, 128);
+
+  stdvga_device_t *stdvga =
+    kmalloc(stdvga_pool, sizeof(stdvga_device_t), M_ZERO);
+
+  uint32_t status_command = pci_read_config(pci, 1, 4);
```

Comment:
I think this line is broken. You can read 4 bytes at once **only** from address aligned to 4.

Please use PCI register names, i.e.:
```c
uint32_t command = pci_read_config(pci, PCIR_COMMAND, 2);
uint32_t status = pci_read_config(pci, PCIR_STATUS, 2);
```